File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.py
class Solution1:
    
    def recur(self, i, prev, nums):
        if i >= len(nums):
            return 0
        
        if self.memo[i][prev+1] != -1:
            return self.memo[i][prev+1]
        
        result = self.recur(i + 1, prev, nums)
        
        if prev == -1 or nums[i] > nums[prev]:
            result = max(result, self.recur(i+1 , i , nums) + 1)
        
        self.memo[i][prev+1] = result
        return result
    
    def lengthOfLIS(self, nums: List[int]) -> int:
        self.memo = [[-1] * (len(nums) + 1) for _ in range(len(nums) + 1)]
        return self.recur(0, -1, nums)
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        
        

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        # Solution - LIS
        # Time - O(N^2)
        # Space - O(N)
        
        dp = [1] * len(nums)

        maxSoFar = 1
        for i in range(1, len(nums)):
            for j in range(i):
                if nums[j] < nums[i]: 
                    dp[i] = max(dp[i],dp[j] + 1)
            maxSoFar = max(dp[i], maxSoFar)

        return maxSoFar
        

File: 300-longest-increasing-subsequence/300-longest-increasing-subsequence.cpp
class Solution {
public:
    vector<int>dp;
    int lengthOfLIS(vector<int>& nums) {
        dp.resize(nums.size(),-1);
        return lis(nums,0,-1);
    }
    
    int lis(vector<int>&nums,int ind,int old_ind){
        if(ind>=nums.size())
            return 0;
        if(dp[old_ind+1]!=-1)
            return dp[old_ind+1];
        int dont_take=lis(nums, ind+1, old_ind);
        int take=1+ lis(nums, ind+1, ind);
        if(old_ind==-1 || nums[ind]>nums[old_ind])
            dp[old_ind+1]=max(take,dont_take);
        else
            dp[old_ind+1]=dont_take;
        return dp[old_ind+1];
    }
};